### ISSUE TITLE
LAB=>RGB conversion

### ISSUE THREAD

**Issue by itaim5** [ZhengyuZhao/PerC-Adversarial#6]
Hi,
It's not an issue but maybe you could help.
I'm doing research on adversarial attacks, and I need both conversions (lab=>rgb, rgb=>lab), so I took your code for rgb=>lab, and implemented the other way around using both your code and other sources (skimage package, wikipedia).
for some reason it didn't work for me. I did notice a fixed per-channel shift in the intermediate xyz representations so once I add the following values [1.6836, 1.7713, 1.9287] after the lab=>xyz conversion (and before the xyz=>rgb) - it works perfectly for most pixels, but still not for the darker pixels.
Adding my code here, maybe you'll see what I did wrong...

`def lab2rgb(lab_image, device):
    y = (lab_image[:,0,:,:] + 16.) / 116.
    x = (lab_image[:,1,:,:] / 500.) + y
    z = y - (lab_image[:,2,:,:] / 200.)

    assert torch.all(z >= 0)

    xyz = torch.stack([x,y,z], dim=1)

    mask1 = xyz > 0.2068966
    mask1_not = torch.logical_not(mask1)

    xyz = (mask1 * xyz) ** 3. + (mask1_not * xyz - 16.0 / 116.) / 7.787

    xyz[:,0,:,:] *= 95.0489
    xyz[:,1,:,:] *= 100.
    xyz[:,2,:,:] *= 108.8840

    # here's the shift I added (have no idea why but it works for most pixels)
    xyz[:,0,:,:] += 1.6836
    xyz[:,1,:,:] += 1.7713
    xyz[:,2,:,:] += 1.9287

    mt = torch.tensor([[0.4124, 0.3576, 0.1805],
                       [0.2126, 0.7152, 0.0722],
                       [0.0193, 0.1192, 0.9504]]).to(device)
    mt_inv = torch.inverse(mt)

    rgb = torch.matmul(mt_inv, xyz.permute(1, 0, 2, 3).contiguous().view(3, -1)).view(3, xyz.size(0),
                                                                                       xyz.size(2),
                                                                                       xyz.size(3)).permute(1, 0, 2, 3)
    rgb /= 100.
    mask2 = rgb > 0.0031308
    mask2_not = torch.logical_not(mask2)
    rgb = 1.055 * (mask2 * rgb) ** (1/2.4) - 0.055 + mask2_not * rgb * 12.92

    return torch.clamp(rgb, 0, 1)`

Thanks in advance :)
Itai

**Comment by ZhengyuZhao** [comment]
Sorry for the reply...

Yes, I am aware that there is actually a small issue in my implementation of color conversion.
Currently, I don't have much time to look into that, but I would suggest that you can check out an alternative here: https://github.com/jeromerony/adversarial-library

Good luck!

**Comment by ZhengyuZhao** [comment]
> Hi, It's not an issue but maybe you could help. I'm doing research on adversarial attacks, and I need both conversions (lab=>rgb, rgb=>lab), so I took your code for rgb=>lab, and implemented the other way around using both your code and other sources (skimage package, wikipedia). for some reason it didn't work for me. I did notice a fixed per-channel shift in the intermediate xyz representations so once I add the following values [1.6836, 1.7713, 1.9287] after the lab=>xyz conversion (and before the xyz=>rgb) - it works perfectly for most pixels, but still not for the darker pixels. Adding my code here, maybe you'll see what I did wrong...
> 
> `def lab2rgb(lab_image, device): y = (lab_image[:,0,:,:] + 16.) / 116. x = (lab_image[:,1,:,:] / 500.) + y z = y - (lab_image[:,2,:,:] / 200.)
> 
> ```
> assert torch.all(z >= 0)
> 
> xyz = torch.stack([x,y,z], dim=1)
> 
> mask1 = xyz > 0.2068966
> mask1_not = torch.logical_not(mask1)
> 
> xyz = (mask1 * xyz) ** 3. + (mask1_not * xyz - 16.0 / 116.) / 7.787
> 
> xyz[:,0,:,:] *= 95.0489
> xyz[:,1,:,:] *= 100.
> xyz[:,2,:,:] *= 108.8840
> 
> # here's the shift I added (have no idea why but it works for most pixels)
> xyz[:,0,:,:] += 1.6836
> xyz[:,1,:,:] += 1.7713
> xyz[:,2,:,:] += 1.9287
> 
> mt = torch.tensor([[0.4124, 0.3576, 0.1805],
>                    [0.2126, 0.7152, 0.0722],
>                    [0.0193, 0.1192, 0.9504]]).to(device)
> mt_inv = torch.inverse(mt)
> 
> rgb = torch.matmul(mt_inv, xyz.permute(1, 0, 2, 3).contiguous().view(3, -1)).view(3, xyz.size(0),
>                                                                                    xyz.size(2),
>                                                                                    xyz.size(3)).permute(1, 0, 2, 3)
> rgb /= 100.
> mask2 = rgb > 0.0031308
> mask2_not = torch.logical_not(mask2)
> rgb = 1.055 * (mask2 * rgb) ** (1/2.4) - 0.055 + mask2_not * rgb * 12.92
> 
> return torch.clamp(rgb, 0, 1)`
> ```
> 
> Thanks in advance :) Itai

Sorry for the late reply...

Yes, I am aware that there is actually a small issue in my implementation of color conversion.
Currently, I don't have much time to look into that, but I would suggest that you can check out an alternative here: https://github.com/jeromerony/adversarial-library

Good luck!
